objects/src/assets/nonfungible.rs

```diff
@@ -68,31 +73,36 @@ impl NonFungibleAsset {
+    pub fn from_parts(faucet_id: AccountIdPrefix, mut data_hash: Word) -> Result<Self, AssetError> {
         if !matches!(faucet_id.account_type(), AccountType::NonFungibleFaucet) {
             return Err(AssetError::NonFungibleFaucetIdTypeMismatch(faucet_id));
         }
         data_hash[FAUCET_ID_POS] = faucet_id.into();
 
-        // Forces the bit at position `ACCOUNT_ISFAUCET_MASK` to `0`.
+        // Forces the bit at position `AccountId::IS_FAUCET_MASK` to `0`.
         //
         // Explanation of the bit flip:
         //
-        // - assets require a faucet account, the id of such accounts always has the bit at the mask
-        //   position.
-        // - fungible assets have the account id at position `3`, meaning the 3rd bit is always of
-        //   the element at the 3rd position is always 1.
-        // - non-fungible assets, have the account id at position `FAUCET_ID_POS`, so the bit at
-        //   position `3` can be used to identify fungible vs. non-fungible assets
+        // - We need to be able to determine the type of an asset by looking at some part of the
+        //   word layout.
+        // - Fungible assets have the first felt of an account id at word index 3. This means the
+        //   bit at the mask position of fungible_asset_word[3] is always 1.
+        // - Non-fungible assets have a data hash at word index 3 and we can set the bit at the mask
+        //   position to 0 explicitly.
+        // - This is necessary because non-fungible assets have the account id prefix at another
+        //   index.
+        // - This means that when looking at the bit at the mask position of an asset, fungible
+        //   assets will have a 1 bit and non-fungible assets will have a 0 bit.
         //
-        // This is done as an optimization, since the field element at position `3` is used as index
-        // when storing the assets into the asset vault. This strategy forces fungible assets to be
-        // assigned to the same slot because it uses the faucet's account id, and allows for easy
-        // merging of fungible faucets. At the same time, it spreads the non-fungible assets evenly
-        // across the vault, because in this case the element is the result of a cryptographic hash
-        // function.
-        let d3 = data_hash[3].as_int();
-        data_hash[3] = Felt::new((d3 & ACCOUNT_ISFAUCET_MASK) ^ d3);
+        // This is done as an optimization, since the field element at position `3` is used as the
+        // index when storing the assets into the asset vault. This strategy forces fungible
+        // assets to be assigned to the same slot because it uses the faucet's account id,
+        // and allows for easy merging of fungible assets. At the same time, it spreads the
+        // non-fungible assets evenly across the vault, because in this case the element is
+        // the result of a cryptographic hash function.
```

Comment:
I'm wondering if we actually need this mechanism. Maybe it would be easier to always store first felt of account ID in position 3 and build vault key for non-fungible assets differently at the time when they are inserted into the vault.

Reply:
Yeah, I was wondering about this as well.

Right now, the asset has this layout: `[hash0, faucet_id_hi, hash2, hash3]`.
In `asset_vault::add_non_fungible_asset` the key is built by duplicating the asset, so `hash3` will be used as the leaf index in the SMT.

If we simply change the layout to `[hash0, hash1, hash2, faucet_id_hi]` then the account ID would be used as the leaf index in the SMT. If I recall correctly, this is not what we want because it would put all of a faucet's assets into the same index. So what you're suggesting is to (before `smt::{set, get}`) construct the vault key of a non-fungible asset so it avoids collisions. For example, by swapping `hash0` with `faucet_id_hi` and then setting the "fungible bit" to zero. I'm guessing the latter is still needed to prevent a collision with some fungible faucet ID, even if that is very unlikely.

I agree that if the asset's layout match, then it's simpler to differentiate them without having to do this fungible bit hack generally.

I implemented this now and I think it's a clearer approach.